### PR TITLE
[dynamo][guards-cpp-refactor] DYNAMIC_INDICES guard

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120469
* #120416
* #120359
* #120344
* #120342
* __->__ #120096
* #120093
* #120123
* #120119
* #120091
* #120089
* #120068
* #120067
* #120065
* #120064
* #120062
* #120061
* #120060
* #119833
* #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -1130,6 +1130,63 @@ class NO_TENSOR_ALIASING : public RelationalGuard {
...
+      : LeafGuard(verbose_code_parts),
+        _has_attr(has_attr),
+        _dynamic_indices(dynamic_indices) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    // Make an interned string
+    static PyObject* dynamic_indices_str =
+        PyUnicode_InternFromString("_dynamo_dynamic_indices");
+
+    bool has_attr = PyObject_HasAttr(value, dynamic_indices_str);
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:1159]: I believe one call to `PyObject_GetAttr` and checking for an exception (nullptr) is faster than `PyObject_HasAttr`+`PyObject_GetAttr`.

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -1130,6 +1130,63 @@ class NO_TENSOR_ALIASING : public RelationalGuard {
...
+    //       if hasattr(x, '_dynamo_dynamic_indices') else True)
+    if (!has_attr) {
+      return true;
+    }
+
+    PyObject* indices = PyObject_GetAttr(value, dynamic_indices_str); // new ref
+    static PyObject* issubset_str = PyUnicode_InternFromString("issubset");
+    PyObject* issubset = PyObject_GetAttr(indices, issubset_str); // new ref
+    PyObject* call_result =
+        PyObject_CallOneArg(issubset, _dynamic_indices.ptr()); // new ref
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:1175]: Use `PyObject_CallMethodOneArg`.  I believe we have a 3.8 back port of that in one of the compatibility headers (for py38).

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp (hunk 2 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:1185]: Does it return a borrowed ref or a new ref? In the latest commit, I assumed its a new ref.

File: torch/csrc/dynamo/guards.cpp (hunk 1 above)
Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp]: Done.

File: torch/csrc/dynamo/guards.cpp (hunk 2 above)
Comment (jansel) [reply torch/csrc/dynamo/guards.cpp]: New reference.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**